pkg/activator/handler/context_handler.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2020 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"go.uber.org/zap"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/logging/logkey"
+	"knative.dev/serving/pkg/activator"
+	"knative.dev/serving/pkg/apis/serving/v1alpha1"
+	revisioninformer "knative.dev/serving/pkg/client/injection/informers/serving/v1alpha1/revision"
+	servinglisters "knative.dev/serving/pkg/client/listers/serving/v1alpha1"
+)
+
+type revisionKey struct{}
+type revIDKey struct{}
+
+// NewContextHandler creates a handler that extracts the necessary context from the request
+// and makes it available on the request's context.
+func NewContextHandler(ctx context.Context, next http.Handler) http.Handler {
+	handler := &contextHandler{
+		nextHandler:    next,
+		revisionLister: revisioninformer.Get(ctx).Lister(),
+		logger:         logging.FromContext(ctx),
+	}
+
+	return handler
+}
+
+// contextHandler enriches the request's context with structured data.
+type contextHandler struct {
+	revisionLister servinglisters.RevisionLister
+	logger         *zap.SugaredLogger
+	nextHandler    http.Handler
+}
+
+func (h *contextHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	namespace := r.Header.Get(activator.RevisionHeaderNamespace)
+	name := r.Header.Get(activator.RevisionHeaderName)
+	revID := types.NamespacedName{Namespace: namespace, Name: name}
+	logger := h.logger.With(zap.String(logkey.Key, revID.String()))
+
+	revision, err := h.revisionLister.Revisions(namespace).Get(name)
+	if err != nil {
+		logger.Errorw("Error while getting revision", zap.Error(err))
+		sendError(err, w)
+		return
+	}
+
+	ctx := r.Context()
+	ctx = logging.WithLogger(ctx, logger)
+	ctx = context.WithValue(ctx, revisionKey{}, revision)
+	ctx = context.WithValue(ctx, revIDKey{}, revID)
+
+	h.nextHandler.ServeHTTP(w, r.WithContext(ctx))
+}
+
+func withRevision(ctx context.Context, rev *v1alpha1.Revision) context.Context {
```

Comment:
do we need the `withX` methods?

Reply:
They help in tests so I kept them here

Reply:
move them to the `context_handler_test.go` then?

Reply:
They are used in all handler tests though to „prepare“ the requests. I‘d keep them as-is for easier readability or do you feel very strongly? I‘ll move them tomorrow in that case, though I really don‘t like the package scoped test helpers we sometimes use across files.